R/cal-estimate.R

```diff
@@ -0,0 +1,405 @@
+cal_estimate_logistic_impl <- function(.data,
+                                       truth = NULL,
+                                       estimate = dplyr::starts_with(".pred_"),
+                                       type,
+                                       model,
+                                       method,
+                                       additional_class,
+                                       ...) {
+  truth <- enquo(truth)
+
+  levels <- truth_estimate_map(.data, !!truth, {{ estimate }})
+
+  if (length(levels) == 2) {
+    log_model <- cal_model_impl(
+      .data = .data,
+      truth = !!truth,
+      estimate = levels[[1]],
+      run_model = model,
+      ...
+    )
+
+    res <- as_binary_cal_object(
+      estimate = log_model,
+      levels = levels,
+      truth = !!truth,
+      method = method,
+      additional_class = additional_class
+    )
+  } else {
+    stop_multiclass()
+  }
+
+  res
+}
+
+cal_model_impl <- function(.data, truth, estimate, run_model, ...) {
+  truth <- ensym(truth)
+
+  if (run_model == "logistic_spline") {
+    f_model <- expr(!!truth ~ s(!!estimate, k = 10))
+    init_model <- mgcv::gam(f_model, data = .data, family = "binomial", ...)
+    model <- butcher::butcher(init_model)
+  }
+
+  if (run_model == "glm") {
+    f_model <- expr(!!truth ~ !!estimate)
+    init_model <- glm(f_model, data = .data, family = "binomial", ...)
+    model <- butcher::butcher(init_model)
+  }
+
+  model
+}
+
+#------------------------------ >> Isotonic ------------------------------------
+cal_isoreg_dataframe <- function(.data,
+                                 truth,
+                                 estimate,
+                                 sampled = FALSE,
+                                 ...) {
+  sort_data <- dplyr::arrange(.data, !!estimate)
+
+  if (sampled) {
+    sort_data <- dplyr::slice_sample(
+      .data = sort_data,
+      prop = 1,
+      replace = TRUE
+    )
+  }
+
+  x <- dplyr::pull(sort_data, !!estimate)
+
+  truth <- dplyr::pull(sort_data, {{ truth }})
+  y <- as.integer(as.integer(truth) == 1)
+
+  model <- isoreg(x = x, y = y)
+
+  model_stepfun <- as.stepfun(model, ...)
+
+  dplyr::tibble(
+    .estimate = environment(model_stepfun)$x,
+    .adj_estimate = environment(model_stepfun)$y
+  )
+}
+
+cal_isoreg_boot <- function(.data,
+                            truth,
+                            estimate,
+                            times,
+                            ...) {
+  sample.int(10000, times) %>%
+    purrr::map(
+      ~ boot_iso(
+        .data = .data,
+        truth = {{ truth }},
+        estimate = {{ estimate }},
+        seed = .x
+      )
+    ) %>%
+    boot_iso_cal()
+}
+
+boot_iso <- function(.data, truth, estimate, seed) {
+  withr::with_seed(
+    seed,
+    {
+      cal_isoreg_dataframe(
+        .data = .data,
+        truth = {{ truth }},
+        estimate = estimate,
+        sampled = TRUE
+      )
+    }
+  )
+}
+
+boot_iso_cal <- function(x) {
```

Comment:
Are we using this? I think that we should use a `distinct()` on the actual data so that we have all possible values without defining a grid. No grids.

Reply:
Again, I think that `findIntervals()` should be used.

Reply:
Ok, I'll be glad to change it. So, where would this code fit? 

```
#' @export
predict.cal_isotonic_boot <- function(object, new_data, ...) {
  cls_nm <- paste0(".pred_", names(object$models[[1]]$levels))
  purrr::map_dfr(object$models, ~ predict(.x, new_data) %>% add_rowindex()) %>%
    setNames(c("prob", "prob2", ".row")) %>%
    group_by(.row) %>%
    summarize(prob = mean(prob, na.rm = TRUE), .groups = "drop") %>%
    mutate(prob2 = 1 - prob) %>%
    arrange(.row) %>%
    select(-.row) %>%
    setNames(cls_nm)
}
```

Reply:
Save the number of `times` models into the calibration object, and then use the average of the predictions for `cal_apply()`